### PR TITLE
MINIFICPP-472 Added date formatting EL functions

### PR DESCRIPTION
**Pull request by achristianson**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by apiri** [comment]
@achristianson Started reviewing.  Build failed on my local envs and it looks like Travis is reproducing the same.  Also seems to be present for #325.  Could you fix those up and rebase please?

**Comment by achristianson** [comment]
Taking a look.

**Comment by apiri** [comment]
Looks like there are still build issues:
https://api.travis-ci.org/v3/job/379278498/log.txt

https://travis-ci.org/apiri/nifi-minifi-cpp/builds/379278497

**Comment by achristianson** [comment]
Ah, seems to be in the 3rd party date lib. We might need to disable these date fns when an old compiler is in use (similar to how we disable regex).

The lib should support C++11 though, so will look into it a bit more before making the call.

**Comment by achristianson** [comment]
@apiri looks like travis issue is fixed. Did you see any other issues?

**Comment by apiri** [comment]
@achristianson will scope out in a bit

File: extensions/expression-language/impl/expression/Expression.h [hunk 1]
```diff
@@ -25,6 +25,11 @@
 #undef EXPRESSION_LANGUAGE_USE_REGEX
 #endif
 
+// Disable date in EL for incompatible compilers
```
Comment (apiri) [inline extensions/expression-language/impl/expression/Expression.h]: Need to define EXPRESSION_LANGUAGE_USE_DATE

**Comment by achristianson** [comment]
@apiri done.

**Comment by apiri** [comment]
Looks like we have build issues with that change.  I attempted building in an Ubuntu 17.10 container:

```
ExpressionLanguageTests is a Catch v1.6.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
Parse Date
-------------------------------------------------------------------------------
/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1216
...............................................................................

/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1221: FAILED:
  REQUIRE( "1398841200000" == expr({flow_file_a}).asString() )
due to unexpected exception with message:
  discover_tz_dir failed to find zoneinfo


-------------------------------------------------------------------------------
Format Date
-------------------------------------------------------------------------------
/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1224
...............................................................................

/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1229: FAILED:
  REQUIRE( "03-14-2014" == expr({flow_file_a}).asString() )
due to unexpected exception with message:
  discover_tz_dir failed to find zoneinfo


-------------------------------------------------------------------------------
Reformat Date
-------------------------------------------------------------------------------
/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1232
...............................................................................

/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1237: FAILED:
  REQUIRE( "03-13-2014" == expr({flow_file_a}).asString() )
due to unexpected exception with message:
  discover_tz_dir failed to find zoneinfo


-------------------------------------------------------------------------------
Now Date
-------------------------------------------------------------------------------
/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1240
...............................................................................

/root/nifi-minifi-cpp/libminifi/test/expression-language-tests/ExpressionLanguageTests.cpp:1249: FAILED:
  REQUIRE( (lt.tm_year + 1900) == expr({flow_file_a}).asUnsignedLong() )
due to unexpected exception with message:
  discover_tz_dir failed to find zoneinfo

```

**Comment by achristianson** [comment]
@apiri I think that might be more of a container issue vs. a build issue. The date lib interacts with the system zoneinfo db, so I think trouble with that is why it is raising a runtime exception during the test.

**Comment by apiri** [comment]
@achristianson ah, fair point.  otherwise looks good to go.  thanks for the fixes!

**Pull request closed without merging**